src/ZendeskApi.Client/Resources/PaginatedResource.cs

```diff
@@ -0,0 +1,24 @@
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using ZendeskApi.Client.Resources.Interfaces;
+
+namespace ZendeskApi.Client.Resources
+{
+    public class PaginatedResource : AbstractBaseResource<PaginatedResource>, IPaginatedResource
```

Comment:
Just trying to wrap my head around this resource. I can see that this is only used in one place: in the `private async Task ExecuteRequest(string requestUrl)` method of the iterator class.

To my mind, these resource classes (and by convention anything implementing `AbstractBaseResource`) represent resources in Zendesk. This does not represent a particular resource and appears to be a helper method of sorts. I can see why you have gone this route - I assume to keep all HTTP requests performed by resources. This does carry weight, but s jeopardises Liskov Substitution Principle. What makes this most prevalent is the fact that the base class enforces the `docsResource` parameter in the base to be present, but this breaks that contract. 

In terms of an alternatives I can think off off the bat...

* I think you just want access to that `ApiClient` object... and might be achieved via Dependency Injection rather than inheritance. I think you can just inject `IZendeskApiClient` into a class which creates the iterator.

* One option is to use a factory to create `CursorPaginatorIterator`. In this factory you could inject the ApiClient dependency needed to make the request.

* At the very least we could make the `PaginatedResource` internally visible because it is only intended for internal use. If we go down that route then a better option might be to have a internal class (not a resource) within the project which exposes the Client object you need. That way it can be injected internally within the package to be used for adhoc HTTP requests like this.

* Not sure what it might look like but a bigger change would be to make the response type from our CBP endpoints to be `CursorPaginatedIterator<T>` instead of `ICursorPagination<T>`. That way we remove the need for consumer to translate from one to the other and just give them the useful thing back

Reply:
thanks @mikerogers123 🙏 

what do you think of my last commit just pushed? I completely removed the the `PaginatedResource` . I agree it didn't make much sense. I switched dependancies on the `CursorPaginationIterator` class to now receive an ApiClient rather than a ZendeskClient. This way I can directly call the method I want. It slightly changes how we do the setup now

```csharp
var client = serviceProvider.GetRequiredService<IZendeskClient>();
var apiClient = serviceProvider.GetRequiredService<IZendeskApiClient>();

var ticketCursorResponse = await client.Tickets.GetAllAsync(new CursorPager { Size = 2 });
var iterator = new CursorPaginatedIterator<Ticket>(ticketCursorResponse, apiClient);
```
Hence the changes in the Factory to accommodate the tests. What do you think?

Reply:
thanks @mikerogers123, 
I think your first suggestion would be cool but also a big and breaking change. Since we don't know exactly how many customers are using this SDK, we would much prefer to not change existing behaviour. 

about your second idea, I might be missing something here... would this factory be responsible for creating a `IZendeskApiClient` to pass to the `CursorPaginatedIterator` ? 

another idea, could we perhaps make a new method in the `ZendeskClient.cs` in order to execute the request? 
This class only has Resources so I'm not sure we would be breaking any sort of design pattern here.

Reply:
🙏

Reply:
Let me know if there's anything else we could work on. it would be great to finish off this work :)

Reply:
thanks @mikerogers123 , can you help me understand how this works in C#? All the examples I see online just have the service required passed as an argument for the class constructor and it's magically injected. 

I have pushed a `wip` commit. Is this more or less what you have in mind? in my case
```cs
public CursorPaginatedIteratorFactory(IZendeskApiClient apiClient)
        {
            ApiClient = apiClient;
        }
```
`ApiClient` is always `null` 

what am I missing, here? 🙏 


... and the code trying to use the SDK 
```cs
// The purpose of this code is to ilustrate how cursor based pagination can be used.
var services = new ServiceCollection();
services.AddZendeskClientWithHttpClientFactory("https://z3n___.zendesk.com", "api-client-r@zendesk.com", "tLzGyE9Wx");
var serviceProvider = services.BuildServiceProvider();
var client = serviceProvider.GetRequiredService<IZendeskClient>();

var ticketCursorResponse = await client.Tickets.GetAllAsync(new CursorPager { Size = 5 });
var iterator = CursorPaginatedIteratorFactory<Ticket>.GetPaginatedIterator(ticketCursorResponse);
```

thanks in advance

Reply:
my wip commit, adding the factory https://github.com/justeat/ZendeskApiClient/pull/322/commits/8d03f2ecb2cdc2b144d65763c1f91594dfd90a79